### DISCUSSION TITLE
Simplify installation custom components / support installation platform

### DISCUSSION THREAD

**Discussion by aex351** [home-assistant/architecture#536] [General]
Home Assistant offers a lot of build-in integrations. However, what makes Home Assistant great is that there is the ability to customize and add you own custom integrations. While apparently Home Assistant has some build-in logic for custom components. There is no means to actually make use of this out-of-the-box and thus you end up installing HACS, which is a custom integration by itself. Most if not all tutorials on Home Assistant actually cover this topic.

It would be great if Home Assistant would offer a platform that would allow:

1. Users to install their own custom components via the interface;
2. Allow users to install a 'custom store' that contains a list of custom components that can be installed.

This does not mean Home Assistant takes responsibility for these custom components or stores. It is just about providing the platform / ability to make use out of it.

To further explain point 2. I see the current list of integrations (https://www.home-assistant.io/integrations/) as the Home Assistant Store.  With this point, any user should be able to draft up their own store, and Home Assistant should be able to 'list' that store as a community/external store alongside the Home Assistant Store.  This would allow home assistant users to create  their own store that consists out of custom components and have the option to share this with other Home Assistant users. Other users simply need to load up that store in Home Assistant.

Furthermore, as a bonus point:
This probably will be a new feature request. It would be great if we can have 'apps'. Which are back-end component + front-end component bundled together. And then also have it list as such in the store. This means for example you can have a 'train app' that users just need to install and without too much additional configuration works out of the box. Now typically users need to install a back-end component (sensor of some sorts) and then need to have a front-end card that goes along with it. While this works for a lot of things. It does not work for everything. And especially for particular use cases it would be nice to have an 'app'.

**Comment by frenck** [comment]
Is this a feature request or are you building it?

**Reply by aex351** [reply]
Yes, let's first start at the beginning, have a discussion on architectural level before any development work happens.

**Reply by frenck** [reply]
Yes, this is a feature request or yes I'm building it?

**Reply by aex351** [reply]
No, I have not started developing it. I think first having a discussion on this topic would be good, as I mentioned in my earlier reply.

**Reply by frenck** [reply]
What is your architectural concern? As in, HACS is there and works I guess, what makes your implementation different that requires changes to our core?

What do these core changes entail?

As in, in the above set story, it looks like a good user story and idea (maybe 🤷 ), but I'm not sure what architectural change is you are looking for? In the end, everything is an integration by itself (like almost everything). Is there a problem with building your integration?

**Reply by aex351** [reply]
It is not about creating a new integration. That could be a potential solution, however I don't think we should start discussing solutions yet. It is about the concept of allowing users to install and manage custom components and community stores from a default Home Assistant perspective. Currently this is not being offered. 

The outcome of this discussion could be a potential solution, which could be a component (integration). I see you are referring to HACS. HACS is a solution by its own means - concept and idea, a standalone integration which users need to install via SSH/etc. to get it going. It doesn't adres any of the points in this concept. The only similarity is that it ultimately allows users to install custom components and that is where the similarity ends.

To continue on the concept. I think managing custom components and stores should be managed via `config/integrations` (divided between original Home Assistant integrations and custom integrations) and a new section `config/stores` to manage the various available stores that contain integrations. it really needs to integral part of Home Assistant. Allowing users a good and seamless user experience on this topic.

**Reply by frenck** [reply]
You are missing the point of my questioning and this repository.
HACS is an integration. The only difference between that and your suggestion right now is how it is distributed.

How stuff becomes part of core? By someone building an integration and adding it into Home Assistant. There are new integrations added every week, adding various things to Home Assistant.

Everything in HA is an integration. A light component, blueprint, automation, system health, Lovelace itself, and HACS.

So, once added to our core repository, it's part of our distribution and no manual installation is needed.

So, back to my original question: what is the architectural change you need to our core that prevents building such an proposed system?

If you are looking a discussion about the general feature / idea of it, the community forum is a better place.

**Reply by aex351** [reply]
I am following this: https://github.com/home-assistant/architecture/discussions/494

> This repository is here to discuss architectural changes across the organization, including:
> 
> Architectural design in our codebases and applications.
> Standards of any kind.
> Policy changes.

My proposal revolves around changing the way how integrations (including custom ones) are going to be managed and used. I am not talking about embedding HACS in to HA.

You seem to talking in terms of components, which are solutions. However I mean this from an architectural point of view, in terms of a concept not a solution. I think to implement this concept, including the bonus point, it revolves around a lot more changes (depending on the chosen solution) than just creating a single component. I am trying to refrain from going in to a solution discussion, as I really feel the concept should be agreed upon first, and then a discussion on how the solution should look like can follow.

If I list my original points:

- Users to install their own custom components via the interface;
This probably will require modifications on http://homeassistant:8123/config/dashboard. User need to be able to have a way to install the custom integrations. If its a agreed upon, it would be great if these custom integrations can me installed via http://homeassistant:8123/config/integrations. So on this page there needs to be a way to add/install and list these custom integrations. As a bonus to this, I can also imagine a page like http://homeassistant:8123/config/integrations/overview, that displays a tiles list (like the website) of all integrations that can be added. Now there only is an 'add integration' button, which gives a list.  

- Allow users to install a 'custom store' that contains a list of custom components that can be installed.
This probably will require modifications on http://homeassistant:8123/config/dashboard. I can imagine a separate page. http://homeassistant:8123/config/store. To manage the custom stores that hold the custom integrations. 

Then for the bonus point, having the integrations listed as apps. There needs to be a way to mark integrations as being apps (front-end + back-end) / back-end only / front-end only. This means there needs to be a way to identify this. Integrations could for example carry a marker (type=app in the manifest.json) etc.

**Reply by frenck** [reply]
You keep talking about features and make me repeat previous answers. Please use the community forums for a general discussion about this.

Thanks!

**Comment by frenck** [comment]
This is a feature request for the community forums at this point.

**Comment by balloob** [comment]
It's already possible to build your own community store, see HACS as example.

We will never built-in an easy way to add custom components because we cannot ensure the quality of them. Anyone can contribute it to Home Assistant if they want it to be accessible easily to Home Assistant users.

**Reply by aex351** [reply]
Why is it relevant to ensure the quality of third party components? Home Assistant has no responsibility for those third party components, therefore it is irrelevant. This is about providing a base so third party components can be installed easily. 

HACS is not a good example, as its pretty much a hack, as I believe it goes beyond what a normal integration can do. However I might be incorrect on that. In any case installing it can be a challenge, and removing it requires manual actions to Home Assistant folders. That is the current user experience people get with installing third party components. 

So if one would make HACS, or something similar, in to a PR so it gets to be part of the Home Assistant base in the vorm of an integration, that would be accepted?

**Reply by frenck** [reply]
> HACS is not a good example, as its pretty much a hack, as I believe it goes beyond what a normal integration can do.

It isn't, it's an integration.

> So if one would make HACS, or something similar, in to a PR so it gets to be part of the Home Assistant base in the vorm of an integration, that would be accepted?

That was answered by balloob above, no it won't.

**Reply by aex351** [reply]
Well it sounds very contradicting and making it unnecessary hard for users. I believe most or at least a lot of users are using HACS. So why block by any means the ability for users to easily install custom components. If one would contribute this as an integration to the base of Home Assistant, then it at least that integration will be maintained. It has got nothing to do with the actual third party components that users install. 

Nb. the integration page even contains this message/disclaimer "Support for these integrations is provided by the Home Assistant community." So there you go, even for the integrations that are currently part of Home Assistant the responsibility is waivered.

**Reply by frenck** [reply]
> Well it sounds very contradicting and making it unnecessary hard for users.

Doesn't have to be, PRs to add them to the core are welcome.

**Reply by aex351** [reply]
No, there is selectiveness in what is and what is not welcome in these statements. If I would make an integration that would allow users to install custom components, why would that get blocked then? Clearly those are not welcome. However, make any random integration that doesn't provide users this ability and it is welcome. Sounds to me there are underlying acceptance criteria to these integrations.

**Reply by frenck** [reply]
Yes, which was answered above.-

**Reply by aex351** [reply]
Great, then it would be good to have this added to the documentation that there are these underlying acceptance criteria. To what an integration can and cannot do. We have one point listed: it is disallowed for an integration to enable installing other integrations.

**Reply by frenck** [reply]
Sure, this probably falls under: modifies or injects code that might change behavior of Home Assistant.

**Reply by aex351** [reply]
Yeah that is good. It will give some clarity to developers, to what is and what is not allowed. 

I am willing to make a PR to add a new page under https://developers.home-assistant.io/docs/development_index `> Building Integration > Acceptance Criteria for Home Assistant integration` to list such criteria.  That would be ok?

**Comment by frenck** [comment]
This proposal has been lingering around. HACS has been the defacto standard for all of this. Our developer documentation in general extended a lot of the last year.

I'm going to close this one for now.

../Frenck
